### PR TITLE
[BEAM-7157] Allow creation of BinaryCombineFn from lambdas.

### PR DESCRIPTION
**Pull request by robertwb**
This came up in a recent user thread.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @youngoli

**Comment by robertwb** [comment]
Run Java PreCommit

**Comment by robertwb** [comment]
I keep getting timeouts in org.apache.beam.runners.dataflow.worker.fn.BeamFnControlServiceTest.testClientConnecting which are irrelevant to this change.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableBiFunction.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * A union of the {@link BiFunction} and {@link Serializable} interfaces.
+ *
+ * @param <FirstInputT> first input value type
+ * @param <SecondInputT> second input value type
+ * @param <OutputT> output value type
+ */
+@FunctionalInterface
+public interface SerializableBiFunction<FirstInputT, SecondInputT, OutputT>
+    extends BiFunction<FirstInputT, SecondInputT, OutputT>, Serializable {
+  // This class is empty, but required for creating serializable lambda expressions.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableBiFunction.java]:
```suggestion
  // This class is empty, but required for creating serializable lambda expression
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java [hunk 2]
```diff
@@ -435,6 +488,27 @@ public OutputT defaultValue() {
...
+    public static <V> BinaryCombineFn<V> of(SerializableBiFunction<V, V, V> combiner) {
+      return new BinaryCombineFn<V>() {
+        @Override
+        public V apply(V left, V right) {
+          return combiner.apply(left, right);
+        }
+      };
+    }
+
+    public static BinaryCombineFn<Long> foo() {
```
Comment (youngoli) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: Are these two functions (foo and bar) supposed to be here? Looks like testing code that was left behind.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableBiFunction.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableBiFunction.java]: Thanks. Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java (hunk 2 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: Oh, yes, I was validating what inference could and could not accept. It's now in the test file but I forgot to delete it here. Thanks.

**Comment by youngoli** [comment]
Run Java PreCommit

**Comment by youngoli** [comment]
testClientConnecting is still failing. I see some precommits are passing. I can't tell if it's flaky or if the fix is in master and isn't included in this PR. I think you should try to get the Java Precommit passing if you can, but the change itself looks good so I'm approving it.

**Comment by robertwb** [comment]
Run Java PreCommit

**Comment by robertwb** [comment]
Thanks for the review. 

Looks like this precommit is now failing more often than not, and spot checking reveals this test to be the culprit. This change is clearly unrelated so I'm going to go ahead and merge.

**Pull request merged**